### ISSUE TITLE
Won't allow download path outside of /var/snap/sabnzbd

### ISSUE THREAD

**Issue by TGP389** [sabnzbd/sabnzbd#2888]
### SABnzbd version

4.3

### Operating system

Ubuntu 24.04

### Using Docker image

None

### Description

I'm trying to set up a new install of sabnzbd on my computer after installing Ubuntu 24.04. I've run sabnzbd both under windows and Ubuntu earlier versions (18 and 22) and had no problem directing the downloads wherever I wished. Now, the browse button on advanced settings download path just shows options in the folder of /var/snap/sabnzbd/. I cannot direct it outside that path. I'm a reformed windows user, and don't know tons about ubuntu, but I tried specifying what I thought would be the absolute path /home/(my user name here)/Downloads/Complete and another with the last folder "Incomplete". My efforts just got me the error message "cannot create..." etc. I tried various combinations, but nothing works, and I can't find any solution anywhere else on the internet. I tried a small download, and it directed it to the indicated default path, but when I tried to move or delete the file, I was disallowed, I suppose because it was outside of the "home" file tree. I believe it's probably something in the newish operating system version, but I'm way past my pay grade here.

**Comment by thezoggy** [comment]
Snap by default is a sandbox, think of it like a vm/container where you can only do stuff within it. You have to use the snap stuff to add mount to be exposed/access to be able to use it. 

Google should tell ya what commands to run, or search for snap in discussion/issues about previous conversation about it

**Comment by TGP389** [comment]
Thanks for your help. I'll search for commands to mount snap.

**Comment by TGP389** [comment]
I must add, though, that I still don't understand why I lost the ability to choose an alternate download path. It's baffling.

**Comment by Safihre** [comment]
Because before you probably used the PPA (see our Ubuntu installation guide on the website). While now you used snap, which is very restrictive.

**Comment by sanderjo** [comment]
@Safihre 

Not your favourite topic, but:

My pycharm is running as snap, and can access my home directory. 

My ksnip is running as snap, and can access & save in my home directory. (FWIW: When I try to save in /tmp, it lands elsewhere: /tmp/snap-private-tmp/snap.ksnip/tmp/ )

Conclusion: snaps can access home.

So how is this done? My guess: in snapcraft.yaml (from https://github.com/ksnip/ksnip/blob/master/snap/snapcraft.yaml )

```
apps:
  ksnip:
...
    plugs:
      - home
```

See https://snapcraft.io/docs/interface-management#:~:text=VLC%20can%20access%20the%20user%E2%80%99s%20/home%20directory%20because%20the%20home%20interface%20is%20connected%20to%20the%20system%20%24HOME%20directory%20(denoted%20by%20the%20%3Ahome%20slot%20name).

So if true, SABnzbd should be able to $home. Let's check sabnzbd snap's connections:

```
$ snap connections sabnzbd 
Interface        Plug                     Slot           Notes
home             sabnzbd:home             :home          -
network          sabnzbd:network          :network       -
network-bind     sabnzbd:network-bind     :network-bind  -
removable-media  sabnzbd:removable-media  -             
``` 

Let's compare:

```
sander@zwart2204:~$ snap connections ksnip | grep home

home                                  ksnip:home                     :home                                                 -

sander@zwart2204:~$ snap connections sabnzbd | grep home

home             sabnzbd:home             :home          -

```

Hmmm ... same, same, butabnzbd seems to have access to home aka $HOME configured, but it's not working?

**Comment by sanderjo** [comment]
Ah, is this a clue:

```
sander@zwart2204:~$ ps -ef | grep -e ksnip -e sabnzbd -e pycharm | cut -c-140
sander     17563    4502  0 13:07 ?        00:00:09 /snap/ksnip/488/bin/ksnip
sander     38736    4502  2 21:18 ?        00:01:15 /snap/pycharm-community/388/jbr/bin/java -classpath /snap/pycharm-community/388/lib/plat
sander     38854   38736  0 21:18 ?        00:00:00 /snap/pycharm-community/388/bin/fsnotifier
root       39380       1  0 21:22 ?        00:00:00 /bin/sh /snap/sabnzbd/6023/bin/sabnzbd-wrapper
root       39408   39380  0 21:22 ?        00:00:03 python3 /snap/sabnzbd/6023/opt/sabnzbd/SABnzbd.py -f /var/snap/sabnzbd/common
sander     42438   33068  0 22:01 pts/1    00:00:00 grep --color=auto -e ksnip -e sabnzbd -e pycharm

```

So the sabnzbd snap is running, whereas ksnip and pycharm are running as sander.

Let's check: yes, SABnzbd can save to `/root/Downloads/incomplete`

```
sander@zwart2204:~$ sudo ls -al /root/Downloads/ | grep -i incomplete
drwxr-xr-x  2 root root 4096 jun 23 22:02 incomplete
```

So $home of root.

Now: how to run the sabnzbd snap as a normal user?

**Comment by sanderjo** [comment]
Ah, I think I found a workaround:

```
sander@zwart2204:~$ sudo snap remove sabnzbd
[sudo] password for sander: 
sabnzbd removed
```

Install with `--devmode` 
```
sander@zwart2204:~$ sudo snap install sabnzbd --devmode
sabnzbd 4.3.2RC1 from Safihre installed
```

And now I can write to /home/sander. Proof:

```
sander@zwart2204:~$ ll snapsab/
total 32
drwxr-xr-x   3 root   root    4096 jun 23 22:26 ./
drwxr-x--- 111 sander sander 20480 jun 23 22:26 ../
drwxr-xr-x   4 root   root    4096 jun 23 22:26 Downloads/

```

**Comment by TGP389** [comment]
> Because before you probably used the PPA (see our Ubuntu installation guide on the website). While now you used snap, which is very restrictive.

Actually, I've installed at least 2 iterations of sabnzb via the snap-store. Both let me choose to put  dowloads wherever. My last install let me use an external drive as the download path.

**Comment by TGP389** [comment]
> Ah, I think I found a workaround:
> 
> ```
> sander@zwart2204:~$ sudo snap remove sabnzbd
> [sudo] password for sander: 
> sabnzbd removed
> ```
> 
> Install with `--devmode`
> 
> ```
> sander@zwart2204:~$ sudo snap install sabnzbd --devmode
> sabnzbd 4.3.2RC1 from Safihre installed
> ```
> 
> And now I can write to /home/sander. Proof:
> 
> ```
> sander@zwart2204:~$ ll snapsab/
> total 32
> drwxr-xr-x   3 root   root    4096 jun 23 22:26 ./
> drwxr-x--- 111 sander sander 20480 jun 23 22:26 ../
> drwxr-xr-x   4 root   root    4096 jun 23 22:26 Downloads/
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/1273502/342116110-9f9246d7-17da-40db-82a5-315252491e09.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkxNzQ4NzksIm5iZiI6MTcxOTE3NDU3OSwicGF0aCI6Ii8xMjczNTAyLzM0MjExNjExMC05ZjkyNDZkNy0xN2RhLTQwZGItODJhNS0zMTUyNTI0OTFlMDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjNUMjAyOTM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTFiZTVhNjQ3YTBmYzFjNzE0OWI0OTQ3ZDZkNjU2N2ZlYWI2MDZhN2U2MmRiOWYwZmJiM2MzN2Y2Zjg3NDU0MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.dqCR04_PIwLf65zMqnvgf_TWhGzX0kSwXXAqc0G5U24)

Thank you so much for the effort. I'll give my modicum of terminal skills a workout. Surely I can cut and paste with the best of them.

**Comment by TGP389** [comment]
I seem to have partially replicated your success. My test download showed up in my home/Downloads/complete folder. The only problem is, a small padlock logo appears on the "complete" folder name, and I'm still unable to move or delete the file. At this rate, my drive would fill up fast, I'm afraid.

**Comment by TGP389** [comment]
I fixed that part myself. I found an entry online with the advice to use the terminal command "sudo chown -R $USER: $HOME" and it dropped the little padlocks.

**Comment by TGP389** [comment]
Thank you one and all for your help and input. I greatly appreciate it.

**Comment by Safihre** [comment]
Some comments here:
Using `devmode` negates the whole point of using snap, since it allows snaps full access to the system. Why use snap at all then?

The problem is that we run as deamon, there can't be a HOME since we run as a service and not as a user. 
If I change that to run as a user, SABnzbd would only start when the user's session starts. Is that useful?
Thoughts @jcfp?

**Comment by sanderjo** [comment]
> Using `devmode` negates the whole point of using snap, since it allows snaps full access to the system. Why use snap at all then?

It does not negate snap's feature of bringing all necessities (libraries, python packages, executables) in one package. You know: avoiding the Linux dependency hell ;-) Just like sabnzbd.exe is self-contained.
I love the pycharm snap: one install command, and you're done.  

It was Mozilla itself who wanted Firefox as a snap package: that way Mozilla can push directly to all Firefox snap users. No dependency on distro bakers and package creators in between.

> If I change that to run as a user, SABnzbd would only start when the user's session starts. Is that useful?

It depends: some users run sabnzbd as a service, unattended, booting at start. Others run will logged on.

I wonder if we turn snap into a non-service and non-root, if you can just start as user from crontab (@reboot) of something more clever (systemd?).

**Comment by TGP389** [comment]
As a non-code writer with limited linux skills, I love the idea of snap. The problem appears that the author of this snap didn't take into account the difficulty of accessing downloads if they're put into a protected directory owned by root, and if that default set up is not alterable by the unenlightened user. I abhor windows, but I can usually get things done in it. In order to use Sabnzbd, I've been forced to reinstall it on an old machine of mine, just for this program, app, or whatever. I love Sabnzbd, and, as I said in my original post, I've used it for years. I couldn't believe it when I installed the updates and couldn't control where downloads are directed. I still can in the windows system.

If I were more linux-savvy, I could probably install Sabnzbd directly, along with all it's dependencies, but I'm not, and since this version is on snap, it seems no one is interested in authoring a how-to for such an installation.

**Comment by sanderjo** [comment]
FYI & prototype:

I changed snapcraft.yaml in 2 lines, created my own snap sabnzbdnondaemon, and I can run it as normal user, and access my home directory


```
sudo snap install sabnzbdnondaemon_4.3.3Beta1_amd64.snap --dangerous
sudo chown sander:sander /var/snap/sabnzbdnondaemon/common/
/snap/bin/sabnzbdnondaemon.sabnzbd
```

```
$ ps -ef | grep -i sabnzbd
sander     68799   44599  0 01:29 pts/3    00:00:00 /bin/sh /snap/sabnzbdnondaemon/x1/bin/sabnzbd-wrapper
sander     68822   68799  6 01:29 pts/3    00:00:00 python3 /snap/sabnzbdnondaemon/x1/opt/sabnzbd/SABnzbd.py -f /var/snap/sabnzbdnondaemon/common
```

![image](https://github.com/user-attachments/assets/67d362a2-5f8d-4be5-aeaa-37d7387aaf5e)

Location not optimal:
```
$ ll /var/snap/sabnzbdnondaemon/common
total 36
drwxr-xr-x 5 sander sander 4096 jul 12 01:23 ./
drwxr-xr-x 4 root   root   4096 jul 12 01:22 ../
drwxrwxr-x 3 sander sander 4096 jul 12 01:29 admin/
drwxrwxr-x 4 sander sander 4096 jul 12 01:23 Downloads/
drwxrwxr-x 2 sander sander 4096 jul 12 01:23 logs/
-rw------- 1 sander sander 8007 jul 12 01:26 sabnzbd.ini
-rw------- 1 sander sander 8007 jul 12 01:26 sabnzbd.ini.bak
```

**Comment by Safihre** [comment]
@TGP389 I am the creator of this snap. And as described above, I had to choose between running as a service or running as a user. Each with their limitations, due to the snap eco system. Note that it has been like that since the creation of the snap, so we didn't suddenly change that in a later update.

Our other Linux ppa package also runs as a service, so that's why this was chosen like this.

**Comment by TGP389** [comment]
Thank you for your work in the snap creation, even though I disagree with the usability of the final product. I'm certainly no expert in linux or ubuntu, although I wish I were, but to me, the snap is not an answer to anything, since I can't modify or delete the product of downloading. To rid myself of downloads, I have to uninstall the entire snap. 

I stumbled on an answer, though. I found a three command installation method on Sabnzbd.org that allowed me to install it so that it operates as it has historically. I can now use the program on Ubuntu again, something for which I'm very grateful, as I've grown to hate windows products.

I really do appreciate both your work and your time spent commenting on my post. I just can't use your snap.

Regards.
